File: sdks/typescript/src/organization.ts
/**
 * Organization, Roles, Scenes, and Norms (RFC 0030).
 *
 * Helpers for declaring an organization manifest extension and enforcing
 * deontic norms (permissions, obligations, prohibitions) across role-scene
 * pairs.
 */

export type DeonticOperator = 'permitted' | 'obligatory' | 'prohibited';

export interface Role {
  role_id: string;
  display_name: string;
  parent_role_id?: string;
}

export interface Scene {
  scene_id: string;
  display_name: string;
  active_when?: string;
}

export interface Norm {
  norm_id: string;
  operator: DeonticOperator;
  role_id: string;
  scene_id: string;
  action_id: string;
  source_rfc?: string;
}

export interface OrganizationManifest {
  organization_did: string;
  display_name: string;
  roles: Role[];
  scenes: Scene[];
  norms: Norm[];
}

export interface DeonticDecision {
  permitted: boolean;
  applicable_norms: Norm[];
  conflict?: { reason: string; norms: Norm[] };
}

export function evaluate(opts: {
  org: OrganizationManifest;
  roleId: string;
  sceneId: string;
  actionId: string;
}): DeonticDecision {
  const applicable = opts.org.norms.filter(
    (n) => n.role_id === opts.roleId && n.scene_id === opts.sceneId && n.action_id === opts.actionId,
  );
  const prohibitions = applicable.filter((n) => n.operator === 'prohibited');
  const obligations = applicable.filter((n) => n.operator === 'obligatory');
  const permissions = applicable.filter((n) => n.operator === 'permitted');

  if (prohibitions.length && obligations.length) {
    return {
      permitted: false,
      applicable_norms: applicable,
      conflict: { reason: 'obligation conflicts with prohibition', norms: [...prohibitions, ...obligations] },
    };
  }
  if (prohibitions.length) return { permitted: false, applicable_norms: applicable };
  if (obligations.length || permissions.length) return { permitted: true, applicable_norms: applicable };
  return { permitted: false, applicable_norms: [] };
}

export function checkConsistency(org: OrganizationManifest): Array<{ kind: string; norms: Norm[] }> {
  const issues: Array<{ kind: string; norms: Norm[] }> = [];
  const seen = new Map<string, Norm[]>();
  for (const n of org.norms) {
    const k = `${n.role_id}|${n.scene_id}|${n.action_id}`;
    const arr = seen.get(k) || [];
    arr.push(n);
    seen.set(k, arr);
  }
  for (const [, arr] of seen) {
    const ops = new Set(arr.map((n) => n.operator));
    if (ops.has('prohibited') && ops.has('obligatory')) {
      issues.push({ kind: 'conflict_prohibited_obligatory', norms: arr });
    }
  }
  return issues;
}

File: sdks/python/src/oap_sdk/organization.py
"""Organization, roles, scenes, deontic norms (RFC 0030)."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional


@dataclass
class Role:
    role_id: str
    display_name: str
    parent_role_id: Optional[str] = None


@dataclass
class Scene:
    scene_id: str
    display_name: str
    active_when: Optional[str] = None


@dataclass
class Norm:
    norm_id: str
    operator: str  # permitted | obligatory | prohibited
    role_id: str
    scene_id: str
    action_id: str
    source_rfc: Optional[str] = None


@dataclass
class OrganizationManifest:
    organization_did: str
    display_name: str
    roles: List[Role]
    scenes: List[Scene]
    norms: List[Norm]


@dataclass
class DeonticDecision:
    permitted: bool
    applicable_norms: List[Norm]
    conflict: Optional[dict] = None


def evaluate(org: OrganizationManifest, role_id: str, scene_id: str, action_id: str) -> DeonticDecision:
    applicable = [
        n for n in org.norms
        if n.role_id == role_id and n.scene_id == scene_id and n.action_id == action_id
    ]
    prohibitions = [n for n in applicable if n.operator == "prohibited"]
    obligations = [n for n in applicable if n.operator == "obligatory"]
    permissions = [n for n in applicable if n.operator == "permitted"]

    if prohibitions and obligations:
        return DeonticDecision(
            permitted=False,
            applicable_norms=applicable,
            conflict={"reason": "obligation conflicts with prohibition", "norms": prohibitions + obligations},
        )
    if prohibitions:
        return DeonticDecision(permitted=False, applicable_norms=applicable)
    if obligations or permissions:
        return DeonticDecision(permitted=True, applicable_norms=applicable)
    return DeonticDecision(permitted=False, applicable_norms=[])


def check_consistency(org: OrganizationManifest) -> List[dict]:
    issues: List[dict] = []
    seen: dict = {}
    for n in org.norms:
        k = f"{n.role_id}|{n.scene_id}|{n.action_id}"
        seen.setdefault(k, []).append(n)
    for arr in seen.values():
        ops = {n.operator for n in arr}
        if "prohibited" in ops and "obligatory" in ops:
            issues.append({"kind": "conflict_prohibited_obligatory", "norms": arr})
    return issues
